apps/aecore/src/aec_sync.erl

```diff
@@ -756,6 +749,65 @@ do_work_on_sync_task(PeerId, Task, LastResult) ->
             epoch_sync:info("~p aborting sync work against ~p", [self(), PeerId])
     end.
 
+agree_on_height(PeerId, #chain{ blocks = [#chain_block{ hash = TopHash, height = TopHeight } | _] }) ->
+    try
+        LocalHeader = aec_chain:top_header(),
+        LocalHeight = aec_headers:height(LocalHeader),
+        MinHeight   = min(TopHeight, LocalHeight),
+        RemoteHash =
+            case TopHeight == MinHeight of
+                true  -> TopHash;
+                false -> get_header_by_height(PeerId, LocalHeight, TopHash)
+            end,
+        case aec_chain:hash_is_connected_to_genesis(RemoteHash) of
+            true ->
+                {ok, MinHeight, RemoteHash};
+            false ->
+                agree_on_height(PeerId, TopHash, MinHeight, 1)
+        end
+    catch throw:{get_header_by_height, {error, Reason}} ->
+        {error, Reason}
+    end.
+
+%% Height is a height where we disagree, so ask for Height - Step
+agree_on_height(PeerId, RemoteTop, Height, Step) ->
+    NewHeight = max(aec_block_genesis:height(), Height - Step),
+    RHash = get_header_by_height(PeerId, NewHeight, RemoteTop),
+    case aec_chain:hash_is_connected_to_genesis(RHash) of
+        true ->
+            agree_on_height(PeerId, RemoteTop, NewHeight, Height, RHash);
```

Comment:
If Height - Step < 0 (genesis height) then NewHeight == 0. We cannot talk to each other if we do not agree ongenesis height. If, however, other party sends us (malliciously) a different hash RHash then genesis Hash, then we loop for ever.

This is a unlikely attack vector, but one could polute the code a bit by `when Height > 0` as guard. Just give up if the we disagree on height 0.

Reply:
@ThomasArts This is handled in the implementation of `get_header_by_height`? If we are indeed at `aec_chain:genesis_height()` (not necessarily 0) we actually don't ask the other node... We just take the genesis_hash from locally since we know they agree... There is even a comment indicating that!